proto/message.go

```diff
@@ -0,0 +1,261 @@
+	Marshal([]byte) error
+
+	// Unmarshal reads the message from the byte slice passed as argument.
+	Unmarshal([]byte) error
+}
+
+// RawMessage represents a raw protobuf-encoded message.
+type RawMessage []byte
+
+// Size satisfies the Message interface.
+func (m RawMessage) Size() int { return len(m) }
+
+// Marshal satisfies the Message interface.
+func (m RawMessage) Marshal(b []byte) error {
+	copy(b, m)
+	return nil
+}
+
+// Unmarshal satisfies the Message interface.
+func (m *RawMessage) Unmarshal(b []byte) error {
+	*m = make([]byte, len(b))
+	copy(*m, b)
+	return nil
+}
+
+// Rewrite satisfies the Rewriter interface.
+func (m RawMessage) Rewrite(out, _ []byte) ([]byte, error) {
+	return append(out, m...), nil
+}
+
+// FieldNumber represents a protobuf field number.
+type FieldNumber int
+
+func (f FieldNumber) Bool(v bool) RawMessage {
+	var x uint64
+	if v {
+		x = 1
+	}
+	return AppendVarint(nil, f, x)
+}
+
+func (f FieldNumber) Int(v int) RawMessage {
+	return f.Int64(int64(v))
+}
+
+func (f FieldNumber) Int32(v int32) RawMessage {
+	return f.Int64(int64(v))
+}
+
+func (f FieldNumber) Int64(v int64) RawMessage {
+	return AppendVarint(nil, f, uint64(v))
+}
+
+func (f FieldNumber) Uint(v uint) RawMessage {
+	return f.Uint64(uint64(v))
+}
+
+func (f FieldNumber) Uint32(v uint32) RawMessage {
+	return f.Uint64(uint64(v))
+}
+
+func (f FieldNumber) Uint64(v uint64) RawMessage {
+	return AppendVarint(nil, f, v)
+}
+
+func (f FieldNumber) Fixed32(v uint32) RawMessage {
+	return AppendFixed32(nil, f, v)
+}
+
+func (f FieldNumber) Fixed64(v uint64) RawMessage {
+	return AppendFixed64(nil, f, v)
+}
+
+func (f FieldNumber) Float32(v float32) RawMessage {
+	return f.Fixed32(math.Float32bits(v))
+}
+
+func (f FieldNumber) Float64(v float64) RawMessage {
+	return f.Fixed64(math.Float64bits(v))
+}
+
+func (f FieldNumber) String(v string) RawMessage {
+	return f.Bytes([]byte(v))
+}
+
+func (f FieldNumber) Bytes(v []byte) RawMessage {
+	return AppendVarlen(nil, f, v)
+}
+
+// Value constructs a RawMessage for field number f from v.
+func (f FieldNumber) Value(v interface{}) RawMessage {
+	switch x := v.(type) {
+	case bool:
+		return f.Bool(x)
+	case int:
+		return f.Int(x)
+	case int32:
+		return f.Int32(x)
+	case int64:
+		return f.Int64(x)
+	case uint:
+		return f.Uint(x)
+	case uint32:
+		return f.Uint32(x)
+	case uint64:
+		return f.Uint64(x)
+	case float32:
+		return f.Float32(x)
+	case float64:
+		return f.Float64(x)
+	case string:
+		return f.String(x)
+	case []byte:
+		return f.Bytes(x)
+	default:
+		panic("cannot rewrite value of unsupported type")
+	}
+}
+
+// The WireType enumeration represents the different protobuf wire types.
+type WireType int
+
+const (
+	Varint  WireType = 0
+	Fixed64 WireType = 1
+	Varlen  WireType = 2
+	Fixed32 WireType = 5
+)
+
+func (wt WireType) String() string {
+	return wireType(wt).String()
+}
+
+func Append(m RawMessage, f FieldNumber, t WireType, v []byte) RawMessage {
```

Comment:
Should we validate the field number (`>= 0`) and wire type (>= 0 && <= 5) here or elsewhere? You may end up with a corrupted tag otherwise:

```
proto/size.go
22-func sizeOfTag(f fieldNumber, t wireType) int {
23:	return sizeOfVarint(uint64(f)<<3 | uint64(t))
```

Reply:
👍 

I changed the types to be unsigned integers.